doc/contributing/primordials.md

```diff
@@ -563,6 +563,62 @@ SafePromiseAll(array, someFunction); // Same as the above, but more efficient.
 
 </details>
 
+### Async functions
+
+<details>
+
+<summary>Awaiting or returning a Promise within the body of an async function
+         depends on user-mutable <code>%Promise.prototype%</code>
+         properties</summary>
+
+Within an async function body, the expression `await v` is internally expanded
+into code that looks like the following:
+
+```js
+let promise;
+if (isPromise(v) && v.constructor === Promise) {
+  promise = v;
+} else if (typeof v === 'object' && typeof v.then === 'function') {
+  promise = new Promise((resolve, reject) => {
+    v.then(resolve, reject);
+  });
+} else {
+  promise = Promise.resolve(v);
+}
```

Comment:
IIUC you're trying to show a "JS implementation" of https://tc39.es/ecma262/#sec-promise-resolve, but it's very confusing to have the `else Promise.resolve(v)` at the end when the whole operation is a PromiseResolve. I think a more accurate description would be:
```suggestion
let promise;
if (isPromise(v) && Object.is(v.constructor, Promise)) {
  promise = v;
} else {
  promise = new Promise((resolve) => {
    resolve(v);
  });
}
```

Reply:
Understood, but the whole point of the pseudoexample is really to show where the handling short-circuits, and under what conditions the vulnerability occurs. Just saying "look up these properties in order, and stop when you find one", as for the more basic prototype lookups, doesn't adequately describe these operations and doesn't provide enough context for the recommendation imo.

Quite happy to go with something more along the lines of the first suggestion, though.

Reply:
Ok sounds good. I still think we need to demonstrate how this can be exploited, e.g.:
```js
// Internal code
async function internalFunction() {
    const internalPromise = new Promise((r) => {r('expected value')});
    const resolvedValue = await internalPromise;
    return resolvedValue;
}

// Userland
delete Promise.prototype.constructor;
Promise.prototype.then = function then(resolve, reject) {
    resolve('overwritten value');
}

// Internal code
PromisePrototypeThen(internalFunction(), console.log); // Outputs: 'overwritten value'